src/source/Rtp/Codecs/RtpH264Payloader.c

```diff
@@ -80,27 +80,38 @@ STATUS getNextNaluLength(PBYTE nalus, UINT32 nalusLength, PUINT32 pStart, PUINT3
     ENTERS();
 
     STATUS retStatus = STATUS_SUCCESS;
-    UINT32 zeroCount = 0;
-    PBYTE pCurPtr = nalus;
-    UINT32 i;
+    UINT32 zeroCount = 0, offset;
+    BOOL naluFound = FALSE;
 
     CHK(nalus != NULL && pStart != NULL && pNaluLength != NULL, STATUS_NULL_ARG);
 
     // Annex-B Nalu will have 0x000000001 or 0x000001 start code, at most 4 bytes
-    for (i = 0; i < 4 && i < nalusLength && nalus[i] == 0; i++);
-
-    CHK(i < nalusLength && i < 4 && i >= 2 && nalus[i] == 1, STATUS_RTP_INVALID_NALU);
-    *pStart = ++i;
+    for (offset = 0; offset < 4 && offset < nalusLength && nalus[offset] == 0; offset++);
+
+    CHK(offset < nalusLength && offset < 4 && offset >= 2 && nalus[offset] == 1, STATUS_RTP_INVALID_NALU);
+    *pStart = ++offset;
+
+    while(offset < nalusLength) {
+        if (nalus[offset] == 0) {
+            /* Maybe next byte is 1 */
+            offset++;
+        } else if (nalus[offset] == 1) {
+            if (nalus[offset - 1] == 0 && nalus[offset - 2] == 0) {
+                zeroCount = nalus[offset - 3] == 0 ? 3 : 2;
+                naluFound = TRUE;
+                break;
+            }
 
-    for (pCurPtr = nalus + i; i < nalusLength && !(*pCurPtr == 1 && (zeroCount == 2 || zeroCount == 3)); i++, pCurPtr++)
-    {
-        if (*pCurPtr == 0) {
-            zeroCount++;
+            /* The jump is always 3 because of the 1 previously matched.
+             * All the 0's must be after this '1' matched at offset */
+            offset += 3;
         } else {
-            zeroCount = 0;
+
+            /* Can jump 3 bytes forward */
+            offset += 3;
         }
     }
-    *pNaluLength = i - *pStart - (zeroCount >= 2 && zeroCount < 4 && i < nalusLength? zeroCount : 0);
+    *pNaluLength = MIN(offset, nalusLength) - *pStart - (naluFound? zeroCount : 0);
```

Comment:
Perhaps we should also check whether we end up with 2 or 3 consequent 0s - that's not legal

Reply:
in that case we may generate false positives if buffer has trailing 0s.

Reply:
NALus should EPB the trailing zeroes. This said, perhaps we are ok to not check this.